### PR TITLE
Lucene embedding store

### LINKED ISSUE #56: [FEATURE] Create Lucene embedding store
## Is your feature request related to a problem? Please describe.
Create a Lucene embedding store. This is not full featured, since it does not support search, but is a step along the way to hybrid full text and vector search.

## Describe the solution you'd like
Create a Lucene store that works well with the Lucene content retriever.

## Describe alternatives you've considered
None.

## Additional context
Step 1 was adding a Lucene content retriever. This is step 2. Step 3 will be hybrid full text and vector search.

### PR DESCRIPTION
**Pull request by sualeh**
## Issue
Closes #56

## Change
Create a new LangChain4J full-text EmbeddingStore using Apache Lucene. This would persist content in Apache Lucene, preserving TextSegment metadata, and include CL100K token counts and similarity scores. It would not support search, since it is full text embedder, not a true embedding store.

This is the next step towards creating a hybrid full text and vector search.

## General checklist
- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit tests in _all_ modules, and they are all green
- [x] I have manually run all integration tests in the module I have added/changed, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module
- [x] I have added my new module in the root `pom.xml` and `langchain4j-community-bom/pom.xml`

## Checklist for adding new embedding store integration
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration\
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by sualeh** [comment]
@Martin7-1 Please note that this embedding store does not support persisting embeddings yet. This is the next step towards creating a hybrid full text and vector search using Lucene.

**Comment by Martin7-1** [comment]
Could you please moving this module to `embedding-stores` directory?

**Comment by sualeh** [comment]
@Martin7-1 I think it should remain where it is. This module is mainly a content retriever, since the full text search (and later, hybrid search) will happen in the content retriever. The embedding store does not have an API (public methods) to support a full text search. That is, `search(EmbeddingSearchRequest request)` does not support full text search in a standardized way.

I have, however, wrapped the Lucene indexing code into a `EmbeddingStore` to make it easy for users of LangChain4J to use. I have added additional methods to make it easy, such as `add(final Embeddeed embedded)`.

@dliubarskyi  - What are your thoughts?

**Comment by sualeh** [comment]
@Martin7-1 Are you able to merge this change? I need this to start working on the next part, which is hybrid search.

**Comment by Martin7-1** [comment]
@sualeh Sorry for the late reply! I'm travelling for Chinese New Year now, will review it next week!

**Comment by dliubarskyi** [comment]
@Martin7-1 happy Chinese New Year! 🥳

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import dev.langchain4j.rag.content.retriever.lucene.LuceneEmbeddingStore;
+import dev.langchain4j.rag.query.Query;
+import java.util.Collections;
+import java.util.List;
+import org.apache.lucene.store.Directory;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class IndexerTest {
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:20]:
```suggestion
public class LuceneEmbeddingStoreIT {
```

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+        indexer.addAll((List<Embedding>) null);
+
+        List<Content> results = contentRetriever.retrieve(query);
+
+        assertThat(results).hasSize(0);
+    }
+
+    @Test
+    public void addEmbedding() {
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:63]:
```suggestion
    public void should_skip_null_embedding() {
```

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        List<Content> results = contentRetriever.retrieve(query);
+
+        assertThat(results).hasSize(1);
+        assertThat(results.get(0).textSegment().metadata().getString("id")).isNotBlank();
+        assertThat(results.get(0).textSegment().text()).isEqualTo(textSegment.text());
+    }
+
+    @Test
+    public void addAllEmbeddings() {
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:51]:
```suggestion
    public void should_skip_add_all_embeddings() {
```

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        metadata.put("name", name);
+        return metadata;
+    }
+
+    private Directory directory;
+    private LuceneEmbeddingStore indexer;
+    private LuceneContentRetriever contentRetriever;
+
+    @Test
+    public void addAll() {
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:37]:
```suggestion
    public void should_add_text() {
```

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/FullTextSearchTest.java [hunk 5]
```diff
@@ -37,14 +37,14 @@ private static Metadata metadataName(String name) {
     }
 
     private Directory directory;
-    private LuceneIndexer indexer;
+    private LuceneEmbeddingStore indexer;
     private LuceneContentRetriever contentRetriever;
 
     @Test
     public void query1() {
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/FullTextSearchTest.java:44]: Would you like to rename this file to `FullTextSearchIT`, and also make the test method name more meaningful?

File: content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    @Override
+    public void add(String id, Embedding embedding) {
+        if (embedding == null) {
+            return;
+        }
+        add(id, embedding, null);
+    }
+
+    /**
+     * Add content to a Lucene index, including segment including metadata and token count. Ids are
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java:133]:
```suggestion
     * Add content to a Lucene index, including segment, metadata and token count. Ids are
```

File: content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+     * generated if they are null. <br>
+     * IMPORTANT: Token counts are approximate, and do not include metadata.
+     *
+     * @param id Content id, can be null
+     * @param embedding Content embedding, can be null
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java:138]: As `LuceneEmbeddingStore` does not add embedding for now, could you please warning for it if `embedding` is not null?

i.e.

```
if (embedding != null) {
    log.warn("LuceneEmbeddingStore does not support add embedding for now.")
}

File: content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+            if (content != null) {
+                Map<String, Object> metadataMap = content.metadata().toMap();
+                if (metadataMap != null) {
+                    for (Entry<String, Object> entry : metadataMap.entrySet()) {
+                        doc.add(toField(entry));
+                    }
+                }
+            }
+
+            writer.addDocument(doc);
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java:173]: I'm wondering if using `writer.addDocuments` would be better, as for now the `LuceneEmbeddingStore` is inserting data one at a time through a loop

File: content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    @Override
+    public void addAll(List<String> idsArg, List<Embedding> embeddingsArg, List<TextSegment> embeddedArg) {
+
+        int maxSize = maxSize(idsArg, embeddingsArg, embeddedArg);
+
+        List<String> ids = ensureSize(idsArg, maxSize);
+        List<Embedding> embeddings = ensureSize(embeddingsArg, maxSize);
+        List<TextSegment> embedded = ensureSize(embeddedArg, maxSize);
+
+        for (int i = 0; i < maxSize; i++) {
```
Comment (Martin7-1) [inline content-retrievers/langchain4j-community-lucene/src/main/java/dev/langchain4j/rag/content/retriever/lucene/LuceneEmbeddingStore.java:215]: Could this be optimized to `writer.addDocuments`?

**Review by Martin7-1 — COMMENTED** [review]
@sualeh Thank you!

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java (hunk 1 above)
Comment (sualeh) [reply content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:20]: @Martin7-1 I do not want to call it that if it does not extend `EmbeddingStoreIT`. Once I get the embedding store working with hybrid search (that is, implement the `search` method, I will see if I can write a *another* test that extends `EmbeddingStoreIT`.

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java (hunk 2 above)
Comment (sualeh) [reply content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:63]: @Martin7-1 Please leave this as is. Think of this as test-driven development. I will implement this in the next PR, and then it will not skip null embeddings.

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java (hunk 3 above)
Comment (sualeh) [reply content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:51]: @Martin7-1 Please leave this as is. Think of this as test-driven development. I will implement this in the next PR, and add another test case that has a non-null embedding in the same test.

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java (hunk 4 above)
Comment (sualeh) [reply content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/IndexerTest.java:37]: @Martin7-1 Please leave this as is. Think of this as test-driven development. I will implement this in the next PR, and add another test case that has a non-null embedding in the same test.

File: content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/FullTextSearchTest.java (hunk 5 above)
Comment (sualeh) [reply content-retrievers/langchain4j-community-lucene/src/test/java/test/dev/langchain4j/rag/content/retriever/FullTextSearchTest.java:46]: Done. I have used display names to help understand what the test is for, rather than renaming test methods.

**Comment by sualeh** [comment]
@Martin7-1 Happy new year! Please take another look at the updates.

**Review by Martin7-1 — APPROVED** [review]
@sualeh Thank you!
